pkg/heyfil/heyfil.go

```diff
@@ -0,0 +1,180 @@
+// service. If the input string is not a Filecoin SP actor address or a peer id,
+// it will be returned as-is.
+func Translate(s string) (string, error) {
+	return Heyfil{}.Translate(s)
+}
+
+// TranslateAll performs a Translate on all strings in the input slice. If none
+// of the strings can be translated, the input slice is returned as-is.
+func (h Heyfil) TranslateAll(ss []string) ([]string, error) {
+	translated := make([]string, len(ss))
+	var merr error
+	var translatedLk sync.Mutex
+	var wg sync.WaitGroup
+	for ii, s := range ss {
+		if !CanHeyfil(s) {
+			translated[ii] = s
+			continue
+		}
+		wg.Add(1)
+		go func(ii int, s string) {
+			defer wg.Done()
+			res, err := h.Translate(s)
+			if err != nil {
+				translatedLk.Lock()
+				multierr.AppendInto(&merr, err)
+				translatedLk.Unlock()
+				return
+			}
+			translatedLk.Lock()
+			translated[ii] = res
+			translatedLk.Unlock()
+		}(ii, s)
+	}
+	wg.Wait()
+	if merr != nil {
+		return nil, merr
+	}
+	return translated, nil
+}
+
+// Translate will translate an input string to a full multiaddr if the string
+// appears to be a Filecoin SP actor address or a peer id using the Heyfil
+// service. If the input string is not a Filecoin SP actor address or a peer id,
+// it will be returned as-is.
+func (h Heyfil) Translate(s string) (string, error) {
+	if isPeerId(s) {
+		res, err := h.translatePeerId(s)
+		if err != nil {
+			logger.Debugw("failed to translate peer id", "input", s, "err", err)
+			return "", err
+		}
+		return res, nil
+	}
+	if isFilecoinFaddr(s) {
+		res, err := h.translateFaddr(s)
+		if err != nil {
+			logger.Debugw("failed to translate faddr", "input", s, "err", err)
+			return "", err
+		}
+		return res, nil
+	}
+	return s, nil // pass-through
+}
+
+func (h Heyfil) translatePeerId(s string) (string, error) {
+	url := h.endpoint() + "/sp?peerid=" + s
+	logger.Debugw("translating peer id", "input", s, "url", url)
+	got := make([]string, 0)
+	if err := httpGet(url, &got); err != nil {
+		return "", err
+	}
+	if len(got) == 0 {
+		return "", fmt.Errorf("expected at least 1 response, got %d", len(got))
+	}
+	return h.translateFaddr(got[0])
+}
+
+func (h Heyfil) translateFaddr(s string) (string, error) {
+	url := h.endpoint() + "/sp/" + s
+	got := make(map[string]interface{})
+	if err := httpGet(url, &got); err != nil {
+		return "", err
+	}
+	addrInfo, ok := got["addr_info"].(map[string]interface{})
+	if !ok {
+		return "", fmt.Errorf("expected addr_info to be map[string]interface{}, got %T", got["addr_info"])
+	}
+	id, ok := addrInfo["ID"].(string)
+	if !ok || id == "" {
+		return "", fmt.Errorf("expected addr_info.ID to be string")
+	}
+	addrs, ok := addrInfo["Addrs"].([]interface{})
+	if !ok {
+		return "", fmt.Errorf("expected addr_info.Addrs to be []interface{}, got %T", addrInfo["Addrs"])
+	}
+	if len(addrs) == 0 {
+		return "", fmt.Errorf("expected addr_info.Addrs to be non-empty")
+	}
+	addr, ok := addrs[0].(string)
```

Comment:
This is probably flawed. It expects at least one Addr and uses the first. It could potentially use multiple addresses and just treat them as separate providers to try. I'm just not sure what's typically recorded here by SPs.